### DISCUSSION TITLE
Welcome to COLA_pytorch Discussions!

### DISCUSSION THREAD

**Discussion by CVxTz** [CVxTz/COLA_pytorch#3] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by KimiaMohsenian** [comment]
Hello!

I hope you are fine and thanks for the repo!

I am trying to use your implementation but with another dataset. (which is clinical audio files)

I first trained the COLA model with the dataset, using your "encode.py" and "train_encoder.py" modules. 
According to what I understood, now this model should have learned representations of the audios, but I don't know where these representations are saved now.
I want to do clustering using these representations of audios.

Thanks for your time!

**Comment by CVxTz** [comment]
hello,

The representations are not saved, but the trained model is.

You can take inspiration from this code: https://github.com/CVxTz/COLA_pytorch/blob/master/supervised_examples/cnn_genre_classification.py
To build a model that encodes your audio data into a vector, for example something similar to this:
```
class AudioToVector(pl.LightningModule):
    def __init__(self, classes=8, p=0.1):
        super().__init__()
        self.save_hyperparameters()

        self.p = p

        self.do = torch.nn.Dropout(p=self.p)

        self.encoder = Encoder(drop_connect_rate=self.p)

        self.g = torch.nn.Linear(1280, 512)
        self.layer_norm = torch.nn.LayerNorm(normalized_shape=512)

        self.fc1 = torch.nn.Linear(512, 256)
        self.fy = torch.nn.Linear(256, classes)

    def forward(self, x):
        x = self.do(self.encoder(x))

        x = self.do(self.g(x))
        x = self.do(torch.tanh(self.layer_norm(x)))
        return x
```
Then load the weights from your cola model into it and use it to predict vectors for each of your samples that you can then use for clustering.

Hope that helps,

Good luck.

**Reply by KimiaMohsenian** [reply]
Thank you so much!

**Comment by KimiaMohsenian** [comment]
Hello!

As the AudioToVector model and COLA model don't have the same architectures as each other, they don't have the same weight dimensions.
For example, there are "fc1" and "fy" wights in the AudioToVector, but not in COLA, also there is a wight of dimension (512, 512) for the linear layer of COLA, but we don't have this layer in AudioToVector; so when loading the wights of the COLA model in the AudioToVector, how should I handle these mismatches?

Thanks in advance!

**Reply by CVxTz** [reply]
Use "model.load_state_dict(ckpt["state_dict"], strict=False)" with strict=False, it will ignore mismatches in layer names.

**Reply by KimiaMohsenian** [reply]
Thanks!

**Comment by KimiaMohsenian** [comment]
Hello!

As you also mentioned in the summary, COLA approach is applied to the audio domain, meaning that audio clips in the time domain must be segmented to produce samples of (anchor, positive) and distractors.
But what I understand from your implementation is this: the Mel-spectrograms are calculated from the audio files and saved, then file paths of these spectrograms are fed to the AudioDataset class when training the encoder (in the train_encoder.py), so these are the spectrograms that are randomly cropped (from their first dimensions) not the audios.
Would you please tell me why this works?

Thanks a lot!

**Reply by CVxTz** [reply]
Hey,

Mel spectrograms are just another representation of audio so the description still holds. You should take a look at the cola paper.

best,

**Comment by KimiaMohsenian** [comment]
Hello!

I hope you are doing well.

There is a part in the training of COLA, which I cannot understand:

`def training_step(self, x, batch_idx):
        x1, x2 = self(x)

        y = torch.arange(x1.size(0), device=x1.device)

        y_hat = torch.mm(x1, x2.t())

        loss = F.cross_entropy(y_hat, y)

        _, predicted = torch.max(y_hat, 1)
        acc = (predicted == y).double().mean()

        self.log("train_loss", loss)
        self.log("train_acc", acc)

        return loss`
 
According to what I understand, x1 and x2 here are the projections of two segments of the spectrogram of a recording, so according to the paper, they form the pair of (anchor, positive) actually, because they belong to a single recording.
So in this code, y_hat is the calculated similarity between these two segments right? then here are my questions:

1. Why is y=torch.arange(x1.size(0), device=x1.device)? I don't understand why our target should be like this.
2. Where does the pair of (anchor, negative) take part in this code? I don't see where the distractors are obtained for an anchor.

Thanks for your time!

**Reply by CVxTz** [reply]
Hello !
For each sample we have a (positive, anchor), there is no explicit negative given to the model.
For each sample (positive, anchor), the negatives (plural) are all the other anchors in the batch. this is why we have the target as y = torch.arange(x1.size(0), device=x1.device), it means that for each element i of x1 we want to match it with the element i of x2.
You should read the paper again in details, it would all make sense.

best,

Youness
